File: src/adventofcode/2018/day19.clj
(ns adventofcode.2018.day19
  (:require clojure.string))

(defn parse-registers [line] (read-string line))
(defn parse-inst [line]
  (as-> line $
    (read-string (str "[:" $ "]"))
    {:op ($ 0)
     :A ($ 1)
     :B ($ 2)
     :C ($ 3)
     }
    ))

(defn parse-program [[ip-decl & lines]]
  {:ip (read-string (second (clojure.string/split ip-decl #"\s+")))
   :instructions (mapv parse-inst lines)
   })

(defn oprr [f]
  (fn [registers {:keys [A B C]}]
    (assoc registers C (f (registers A) (registers B)))
    ))

(defn opri [f]
  (fn [registers {:keys [A B C]}]
    (assoc registers C (f (registers A) B))
    ))

(defn opir [f]
  (fn [registers {:keys [A B C]}]
    (assoc registers C (f A (registers B)))
    ))

(defn boolf [f] (comp {true 1 false 0} f))

(def ops {
          :addr (oprr +)
          :addi (opri +)
          :mulr (oprr *)
          :muli (opri *)
          :banr (oprr bit-and)
          :bani (opri bit-and)
          :borr (oprr bit-or)
          :bori (opri bit-or)
          :setr (opri (fn [A B] A))
          :seti (fn [registers inst] (assoc registers (:C inst) (:A inst)))
          :gtir (opir (boolf >))
          :gtri (opri (boolf >))
          :gtrr (oprr (boolf >))
          :eqir (opir (boolf =))
          :eqri (opri (boolf =))
          :eqrr (oprr (boolf =))
          })

(defn initial-state []
  {:registers [0 0 0 0 0 0]
   :ip 0
   })

(defn step [program state]
  (let [ip (:ip state)
        inst ((:instructions program) ip)
        ]
    (-> state
        (assoc-in [:registers (:ip program)] ip)
        (update :registers #((ops (:op inst)) % inst))
        (as-> $ (assoc $ :ip ((:registers $) (:ip program))))
        (update :ip inc)
        )))

(defn halted? [program state]
  (not (contains? (:instructions program) (:ip state))))

(defn run-program [program state]
  (if (halted? program state)
    state
    (recur program (step program state))
    ))

(defn divisors [f]
  (->> (range 1 (inc f))
       (keep (fn [b]
               (if (= 0 (mod f b)) b nil)))
       (apply +)
       ))

(defn solve-reverse-engineered [initial-a lines]
  (-> lines
      (parse-program)
      (as-> $ (iterate (partial step $) (assoc-in (initial-state) [:registers 0] initial-a)))
      (as-> $ (filter (fn ([state]
                           (= 1 (:ip state))))
                      $))
      (first)
      (:registers)
      (as-> $ (apply max $))
      (divisors)
  ))

(defn solve-a [lines]
  (solve-reverse-engineered 0 lines))

(defn solve-b [lines]
  (solve-reverse-engineered 1 lines))

(defn run [input-lines & args]
  {:A (solve-a input-lines)
   :B nil
   }
  nil
  )

(defn format-program [program]
  (->> program
       (:instructions)
       (map-indexed (fn [i inst]
                      (str i " " inst)
                      ))
       (clojure.string/join \newline)
       ))

(defn format-state [program state]
  (->> program
       (format-program)
       (clojure.string/split-lines)
       (map-indexed (fn [i line]
                      (if (= i (:ip state))
                        (str line " <--")
                        (str line)
                        )))
       (clojure.string/join \newline)
       (str state \newline)
       ))

(def example-input "#ip 0
seti 5 0 1
seti 6 0 2
addi 0 1 0
addr 1 2 3
setr 1 0 0
seti 8 0 4
seti 9 0 5")
(defn day-lines [] (adventofcode.2018.core/day-lines 19))
(def states [])
(def program nil)
(defn show-state [] (println (format-state program (last states))))
(defn start [lines] (def program (parse-program lines)) (def states [(initial-state)]) (show-state))
(defn start-day-lines [] (start (day-lines)))
(defn start-example [] (start (clojure.string/split-lines example-input)))
(defn n []
  (def states (conj states (step program (last states))))
  (show-state))
(defn p []
  (def states (pop states))
  (show-state))

File: dlofstrom-python/day19.py
import sys
import re

# Read input
input = sys.stdin.read().strip()

input = input.split('\n')
input = [l.split(' ') for l in input]

def addr(A,B,C,registers):
    registers[C] = registers[A] + registers[B]
    return registers
    
def addi(A,B,C,registers):
    registers[C] = registers[A] + B
    return registers

def mulr(A,B,C,registers):
    registers[C] = registers[A] * registers[B]
    return registers

def muli(A,B,C,registers):
    registers[C] = registers[A] * B
    return registers

def banr(A,B,C,registers):
    registers[C] = registers[A] & registers[B]
    return registers

def bani(A,B,C,registers):
    registers[C] = registers[A] & B
    return registers

def borr(A,B,C,registers):
    registers[C] = registers[A] | registers[B]
    return registers

def bori(A,B,C,registers):
    registers[C] = registers[A] | B
    return registers

def setr(A,B,C,registers):
    registers[C] = registers[A]
    return registers

def seti(A,B,C,registers):
    registers[C] = A
    return registers

def gtir(A,B,C,registers):
    if A > registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def gtri(A,B,C,registers):
    if registers[A] > B:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def gtrr(A,B,C,registers):
    if registers[A] > registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def eqir(A,B,C,registers):
    if A == registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def eqri(A,B,C,registers):
    if registers[A] == B:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

def eqrr(A,B,C,registers):
    if registers[A] == registers[B]:
        registers[C] = 1
    else:
        registers[C] = 0
    return registers

oplist = {"addr":addr, "addi":addi, "mulr":mulr, "muli":muli, "banr":banr, "bani":bani, "borr":borr, "bori":bori, "setr":setr, "seti":seti, "gtir":gtir, "gtri":gtri, "gtrr":gtrr, "eqir":eqir, "eqri":eqri, "eqrr":eqrr}

ir = int(input.pop(0)[1])
input = [(o,int(a),int(b),int(c)) for o,a,b,c in input]
ip = 0
registers = [0,0,0,0,0,0]

# Run program
ip = registers[ir]
while ip >= 0 and ip < len(input):
    o,a,b,c = input[ip]
    rb = list(registers)
    registers = oplist[o](a,b,c,list(registers))
    registers[ir] += 1
    ip = registers[ir]

print('Part 1:', registers[0])


# The program finds the sum of divisors of a large number
# Run program enough loops to generate the number
ip = 0
registers = [1,0,0,0,0,0]
ip = registers[ir]
for i in range(100):
    o,a,b,c = input[ip]
    rb = list(registers)
    registers = oplist[o](a,b,c,list(registers))
    registers[ir] += 1
    ip = registers[ir]

# Extract number
n = max(registers)

# Find sum of divisors
ds = []
for i in range(1,n+1):
    if n % i == 0:
       ds.append(i)
print('Part 2:', sum(ds))
